CIPs/0007.md

```diff
@@ -0,0 +1,108 @@
+# CIP [0007]: Affordable Gas Downtime Slasher
+
+- Date: 2020-05-05
+- Author: @gastonponti (with a lot of feedback and ideas from @m-chrzan)
+- Status: WIP
+
+## Overview
+
+This proposal contains a new DowntimeSlasher contract.
+
+The actual DowntimeSlasher tries to verify that the validator signed at least a block in all the window that is required to be validated which it leads to running out of gas for big windows.
+
+The main goal of this new contract is to generate a tool that would allow any user to slash a validator for every possible block windows without gas limitations. 
+
+For the actual window of 1/2 epoch, the DowntimeSlasher won't finish with a "run out gas" exception. Without this new tool for slashing, there is no a possible way to penalize a validator for downtime.
+
+## Goals
+
+Make a new contract that:
+- Allows the user to slash a validator if it was down for a specific consecutive block window
+- The implementation works for every possible block window (required in the initialization)
+
+## Proposed Solution
+
+Create a new contract funtion that divides the downtime verification in valid slots.
+
+The user that wants to slash a particular validator will have to call a function from the contract `S_n` times (`S_n` could be asked to the contract itself) to verify each slot.
+
+Then the user must generate another contract call to verify that the validator to be slashed was down for every consecutive slot.
+
+### Process
+
+1) The user calls the contract to know how many slots are required (`S_n`) and how many blocks must have every slot (`B_n`)
+
+2) The user requires to validate `S_n` slots
+    - The user calls the contract for the slot validation `[StartBlockNumber + (i * B_n), StartBlockNumber + ((1 + i) * B_n) )` with `0 <= i < S_n`
+
+3) The user calls the slash method with the `StartBlockNumber` (internally will verify that every slot validation was performed and succeded)
+
+### Variation of the proposed solution
```

Comment:
Which of these are you proposing?

Reply:
Well, I'm not totally decided, and I want to discuss it a little.

For my point of view, the "user has to verify everything" it's the approach that would be easier for the community to accept because it's not changing anything (same as the old one).
That's why I was thinking to use that. But I would like some opinions on that.

(And you are totally right, I should write which option I'm proposing clearer)

Reply:
I like the 1st proposal.  It's much more intuitive (which will incidentally make implementation much more easier) and I would think that the added functionality of the 2nd proposal will hardly be utilized.  E.g. it will almost always be one one user that generates all of the validation slots.  I suppose that we could also upgrade the smart contract later on down the road if there is alot of community demand for the 2nd proposal.